File: dotnet/02_QueryCall.cs
// Recipe 02: query Calls via /calls/stream_query.
//
// Demonstrates the workhorse read endpoint:
//   POST /calls/stream_query  -> stream NDJSON of matching Calls
//
// Sets up by creating one Call (op_name="recipe-02-query-call"), then
// queries that op_name and confirms the just-created Call appears in
// the streamed results.
//
// The endpoint returns one JSON object per line (application/jsonl).
// We parse line-by-line via StreamReader rather than buffering the
// full response, demonstrating the streaming pattern in HttpClient.
//
// Run:
//   dotnet run dotnet/02_QueryCall.cs

using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization.Metadata;
using System.Threading;
using System.Threading.Tasks;

// .NET 10 file-based programs disable reflection-based JSON serialization
// by default; this resolver re-enables it. See 01_StartCall.cs for context.
JsonSerializerOptions jsonOptions = new() { TypeInfoResolver = new DefaultJsonTypeInfoResolver() };

string baseUrl = Environment.GetEnvironmentVariable("WEAVE_SERVICE_URL") ?? "https://trace.wandb.ai";

string[] required = { "WANDB_API_KEY", "WANDB_ENTITY", "WANDB_PROJECT" };
var missing = new List<string>();
foreach (var k in required)
    if (string.IsNullOrEmpty(Environment.GetEnvironmentVariable(k))) missing.Add(k);
if (missing.Count > 0)
{
    Console.Error.WriteLine($"Missing required env vars: {string.Join(", ", missing)}. See ../README.md#setup.");
    return 1;
}

string apiKey = Environment.GetEnvironmentVariable("WANDB_API_KEY")!;
string projectId = $"{Environment.GetEnvironmentVariable("WANDB_ENTITY")}/{Environment.GetEnvironmentVariable("WANDB_PROJECT")}";

const string opName = "recipe-02-query-call";
var attributes = new Dictionary<string, object>
{
    ["cookbook.language"] = "dotnet",
    ["cookbook.recipe"] = "02_query_call",
    ["cookbook.environment"] = Environment.GetEnvironmentVariable("COOKBOOK_ENVIRONMENT") ?? "dev",
};
var inputs = new Dictionary<string, object> { ["question"] = "What is the capital of Spain?" };
var output = new Dictionary<string, object> { ["answer"] = "Madrid" };

using var http = new HttpClient();
http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(
    "Basic",
    Convert.ToBase64String(Encoding.ASCII.GetBytes($"api:{apiKey}")));

async Task<JsonNode> PostJson(string path, object body)
{
    var json = JsonSerializer.Serialize(body, jsonOptions);
    var res = await http.PostAsync(baseUrl + path, new StringContent(json, Encoding.UTF8, "application/json"));
    var responseBody = await res.Content.ReadAsStringAsync();
    if (!res.IsSuccessStatusCode)
        throw new HttpRequestException($"HTTP {(int)res.StatusCode} for {path}: {responseBody}");
    return string.IsNullOrEmpty(responseBody) ? new JsonObject() : JsonNode.Parse(responseBody)!;
}

// Streams /calls/stream_query response line-by-line, parsing each
// newline-delimited JSON object as it arrives. Returns the parsed rows.
async Task<List<JsonNode>> StreamQuery(object body)
{
    var json = JsonSerializer.Serialize(body, jsonOptions);
    using var req = new HttpRequestMessage(HttpMethod.Post, baseUrl + "/calls/stream_query")
    {
        Content = new StringContent(json, Encoding.UTF8, "application/json"),
    };
    using var res = await http.SendAsync(req, HttpCompletionOption.ResponseHeadersRead);
    if (!res.IsSuccessStatusCode)
    {
        var errBody = await res.Content.ReadAsStringAsync();
        throw new HttpRequestException($"HTTP {(int)res.StatusCode} for /calls/stream_query: {errBody}");
    }
    var rows = new List<JsonNode>();
    using var stream = await res.Content.ReadAsStreamAsync();
    using var reader = new StreamReader(stream);
    string? line;
    while ((line = await reader.ReadLineAsync()) is not null)
    {
        if (string.IsNullOrWhiteSpace(line)) continue;
        rows.Add(JsonNode.Parse(line)!);
    }
    return rows;
}

// Setup: create + end a Call we can later query for.
var started = await PostJson("/call/start", new
{
    start = new
    {
        project_id = projectId,
        op_name = opName,
        started_at = DateTime.UtcNow.ToString("O"),
        attributes,
        inputs,
    },
});
var callId = started["id"]!.GetValue<string>();
var traceId = started["trace_id"]!.GetValue<string>();
Console.WriteLine($"Created: id={callId}");

await PostJson("/call/end", new
{
    end = new
    {
        project_id = projectId,
        id = callId,
        ended_at = DateTime.UtcNow.ToString("O"),
        summary = new { },
        output,
    },
});

// Query: stream Calls matching our op_name, newest first. Retry briefly
// to tolerate eventual consistency on the read path.
JsonNode? found = null;
for (var attempt = 0; attempt < 5; attempt++)
{
    var rows = await StreamQuery(new
    {
        project_id = projectId,
        filter = new { op_names = new[] { opName } },
        sort_by = new[] { new { field = "started_at", direction = "desc" } },
        limit = 50,
    });
    // Require ended_at populated so we don't race the write-to-read
    // propagation and read a half-finalized row.
    found = rows.Find(c => c["id"]?.GetValue<string?>() == callId
                            && c["ended_at"]?.GetValue<string?>() != null);
    if (found != null) break;
    Thread.Sleep(1000);
}

try
{
    if (found is null)
        throw new Exception($"Call {callId} not in stream_query results after 5 attempts");

    if (found["op_name"]!.GetValue<string>() != opName)
        throw new Exception($"op_name: {found["op_name"]}");
    foreach (var (k, v) in attributes)
    {
        var actual = found["attributes"]?[k]?.GetValue<string?>();
        if (actual != v.ToString())
            throw new Exception($"attribute {k}: {actual}");
    }
    if (found["inputs"]?["question"]?.GetValue<string?>() != "What is the capital of Spain?")
        throw new Exception($"inputs: {found["inputs"]}");
    if (found["output"]?["answer"]?.GetValue<string?>() != "Madrid")
        throw new Exception($"output: {found["output"]}");
    if (found["trace_id"]!.GetValue<string>() != traceId)
        throw new Exception($"trace_id: {found["trace_id"]}");

    Console.WriteLine($"Verified: id={callId}");
    return 0;
}
catch (Exception e)
{
    Console.Error.WriteLine($"FAIL: {e.Message}");
    return 1;
}

File: ruby/02_query_call.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Recipe 02: query Calls via /calls/stream_query.
#
# Demonstrates the workhorse read endpoint:
#   POST /calls/stream_query  -> stream NDJSON of matching Calls
#
# Sets up by creating one Call (op_name="recipe-02-query-call"), then
# queries that op_name and confirms the just-created Call appears in
# the streamed results.
#
# The endpoint returns one JSON object per line (application/jsonl).
# We parse line-by-line via a chunk buffer rather than reading the
# whole body, demonstrating the streaming pattern in Net::HTTP.
#
# Run:
#   ruby ruby/02_query_call.rb

require "json"
require "net/http"
require "time"
require "uri"

BASE_URL = ENV.fetch("WEAVE_SERVICE_URL", "https://trace.wandb.ai")

required = %w[WANDB_API_KEY WANDB_ENTITY WANDB_PROJECT]
missing = required.reject { |k| ENV[k] && !ENV[k].empty? }
abort "Missing required env vars: #{missing.join(", ")}. See ../README.md#setup." unless missing.empty?

PROJECT_ID = "#{ENV.fetch("WANDB_ENTITY")}/#{ENV.fetch("WANDB_PROJECT")}"
API_KEY = ENV.fetch("WANDB_API_KEY")

OP_NAME = "recipe-02-query-call"
ATTRIBUTES = {
  "cookbook.language" => "ruby",
  "cookbook.recipe" => "02_query_call",
  "cookbook.environment" => ENV.fetch("COOKBOOK_ENVIRONMENT", "dev"),
}.freeze
INPUTS = { "question" => "What is the capital of Spain?" }.freeze
OUTPUT = { "answer" => "Madrid" }.freeze

def post_json(path, body)
  uri = URI.join(BASE_URL, path)
  req = Net::HTTP::Post.new(uri, "Content-Type" => "application/json")
  req.basic_auth("api", API_KEY)
  req.body = JSON.dump(body)
  res = Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") { |http| http.request(req) }
  abort "HTTP #{res.code} for #{path}: #{res.body}" unless res.code.start_with?("2")
  res.body.empty? ? {} : JSON.parse(res.body)
end

# Streams /calls/stream_query response chunk-by-chunk, parsing each
# newline-delimited JSON object as it arrives. Returns the parsed rows.
def stream_query(body)
  uri = URI.join(BASE_URL, "/calls/stream_query")
  req = Net::HTTP::Post.new(uri, "Content-Type" => "application/json")
  req.basic_auth("api", API_KEY)
  req.body = JSON.dump(body)

  buffer = +""
  rows = []
  Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") do |http|
    http.request(req) do |res|
      abort "HTTP #{res.code} for /calls/stream_query: #{res.read_body}" unless res.code.start_with?("2")
      res.read_body do |chunk|
        buffer << chunk
        while (idx = buffer.index("\n"))
          line = buffer.slice!(0..idx).chomp
          rows << JSON.parse(line) unless line.empty?
        end
      end
    end
  end
  rows << JSON.parse(buffer) unless buffer.strip.empty?
  rows
end

# Setup: create + end a Call we can later query for.
started = post_json("/call/start", {
  "start" => {
    "project_id" => PROJECT_ID,
    "op_name" => OP_NAME,
    "started_at" => Time.now.utc.iso8601,
    "attributes" => ATTRIBUTES,
    "inputs" => INPUTS,
  },
})
call_id = started.fetch("id")
trace_id = started.fetch("trace_id")
puts "Created: id=#{call_id}"

post_json("/call/end", {
  "end" => {
    "project_id" => PROJECT_ID,
    "id" => call_id,
    "ended_at" => Time.now.utc.iso8601,
    "summary" => {},
    "output" => OUTPUT,
  },
})

# Query: stream Calls matching our op_name, newest first. Retry briefly
# to tolerate eventual consistency on the read path.
found = nil
5.times do
  results = stream_query({
    "project_id" => PROJECT_ID,
    "filter" => { "op_names" => [OP_NAME] },
    "sort_by" => [{ "field" => "started_at", "direction" => "desc" }],
    "limit" => 50,
  })
  # Require ended_at populated so we don't race the write-to-read
  # propagation and read a half-finalized row.
  found = results.find { |c| c["id"] == call_id && c["ended_at"] }
  break if found

  sleep 1
end

# --- verification ---
abort "FAIL: Call #{call_id} not in stream_query results after 5 attempts" if found.nil?

abort "op_name mismatch: #{found["op_name"].inspect}" unless found["op_name"] == OP_NAME
ATTRIBUTES.each do |k, v|
  abort "attribute #{k} mismatch: #{found["attributes"][k].inspect}" unless found["attributes"][k] == v
end
abort "inputs mismatch: #{found["inputs"].inspect}" unless found["inputs"] == INPUTS
abort "output mismatch: #{found["output"].inspect}" unless found["output"] == OUTPUT
abort "trace_id mismatch: #{found["trace_id"].inspect}" unless found["trace_id"] == trace_id
puts "Verified: id=#{call_id}"
